ejp_xml_pipeline/etl.py

```
@@ -80,23 +92,110 @@ def etl_ejp_xml_zip(
+def load_entities_file_to_s3(
+        ejp_xml_load_config: eJPXmlDataConfig,
+        original_obj_key
+):
     for entity in ejp_xml_load_config.entity_type_mapping.values():
         if os.path.getsize(entity.get_full_file_location()) > 0:
-            create_or_extend_table_schema(
-                ejp_xml_load_config.gcp_project,
-                ejp_xml_load_config.dataset,
-                entity.table_name,
-                entity.get_full_file_location()
+            obj_key = get_temp_s3_object_name(
+                entity.s3_object_prefix,
+                original_obj_key
+            )
+            upload_file_into_s3(
+                bucket=ejp_xml_load_config.temp_file_s3_bucket,
+                object_key=obj_key,
+                full_file_path=entity.get_full_file_location()
             )
 
-            load_file_into_bq(
-                filename=entity.get_full_file_location(),
-                table_name=entity.table_name,
-                dataset_name=ejp_xml_load_config.dataset,
-                project_name=ejp_xml_load_config.gcp_project
+
+def load_entity_file_to_bq(
+        gcp_project: str,
+        dataset: str,
+        table_name: str,
+        file_path: str
+):
+    if os.path.getsize(file_path) > 0:
+        create_or_extend_table_schema(
+            gcp_project,
+            dataset,
+            table_name,
+            file_path
+        )
+        load_file_into_bq(
+            filename=file_path,
+            table_name=table_name,
+            dataset_name=dataset,
+            project_name=gcp_project
+        )
+
+
+# pylint: disable='too-many-arguments'
+def download_load2bq_cleanup_temp_files(
+        matching_file_metadata_iter, s3_bucket: str,
+        gcp_project: str, dataset: str,
+        bq_table: str, batch_size_limit: int = 5000
+):
+    written_file_row_count = 0
+    s3_objects_written_to_file = []
+    with NamedTemporaryFile() as named_temp_file:
```

I think we had issues with `NamedTemporaryFile` before. I find it a bit more difficult to work with because one needs to be careful that it is still readable / writable, file is flushed and file cursor is at the right location.

in this case I am not sure whether we may need to flush the file passing it on.

---

[bigquery python data load api](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.load_table_from_file) only works with strings, and that is the reason I am using `NameTemporaryFile`. If I know the main issues with using `NamedTemporaryFile`, I can try to to see that this is avoided e.g. using flush like  you suggested.

---

Only saw your response now. I am not sure this is related to the BigQuery API.

The main issue seem to be the need to `flush`. But having to remember that as it would potentially fail silently or not, would be a reason for not using it. I would use [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) and then simple file operations on files within it, closing the file before reading it (via separate context manager). To me that is easier to reason about and the overhead shouldn't be noticeable for most use cases.

---

I think additionally is the complexity that `NamedTemporaryFile` will already open the file. It is then opened again, which may even be rejected on Windows (we don't target that but it seems simple to avoid it):

> Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later).

I guess in this case the flush is not necessary as `named_temp_file` is not written to. I think one shouldn't have to think about it for general use.